lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py

```diff
@@ -0,0 +1,217 @@
+     required: true
+   url:
+     description:
+        - URL of the service.
+     required: true
+   region:
+     description:
+        - Region that the service belongs to. Note that I(region_name) is used for authentication.
+   enabled:
+     description:
+        - Is the service enabled.
+     default: True
+   state:
+     description:
+       - Should the resource be C(present) or C(absent).
+     choices: [present, absent]
+     default: present
+requirements:
+    - shade >= 1.11.0
+'''
+
+EXAMPLES = '''
+- name: Create a service for glance
+  os_keystone_endpoint:
+     cloud: mycloud
+     service: glance
+     interface: public
+     url: http://controller:9292
+     region: RegionOne
+     state: present
+
+- name: Delete a service for nova
+  os_keystone_endpoint:
+     cloud: mycloud
+     service: nova
+     interface: public
+     region: RegionOne
+     state: absent
+'''
+
+RETURN = '''
+endpoint:
+    description: Dictionary describing the endpoint.
+    returned: On success when I(state) is C(present)
+    type: complex
+    contains:
+        id:
+            description: Endpoint ID.
+            type: string
+            sample: 3292f020780b4d5baf27ff7e1d224c44
+        region:
+            description: Region Name.
+            type: string
+            sample: RegionOne
+        service_id:
+            description: Service ID.
+            type: string
+            sample: b91f1318f735494a825a55388ee118f3
+        interface:
+            description: Endpoint Interface.
+            type: string
+            sample: public
+        url:
+            description: Service URL.
+            type: string
+            sample: http://controller:9292
+        enabled:
+            description: Service status.
+            type: boolean
+            sample: True
+'''
+
+try:
+    import shade
+    HAS_SHADE = True
+except ImportError:
+    HAS_SHADE = False
+
+from distutils.version import StrictVersion
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.openstack import openstack_full_argument_spec
+from ansible.module_utils.openstack import openstack_module_kwargs
+
+
+def _needs_update(module, endpoint):
+    if endpoint.enabled != module.params['enabled']:
+        return True
+    if endpoint.url != module.params['url']:
+        return True
+    return False
+
+
+def _system_state_change(module, endpoint):
+    state = module.params['state']
+    if state == 'absent' and endpoint:
+        return True
+
+    if state == 'present':
+        if endpoint is None:
+            return True
+        return _needs_update(module, endpoint)
+
+    return False
+
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        service=dict(type='str', required=True),
+        interface=dict(type='str', required=True, choices=['admin', 'public', 'internal']),
+        url=dict(type='str', required=True),
+        region=dict(type='str', required=False, default=None),
```

Comment:
So we tend not to specify `required=False` and `default=None` because that is the case by default.
(Although we still like to state the type, even the default nowadays is "str")

Some of these rules are codified, I would like the sanity tests to enforce some of this.

Reply:
So to give you some reasons why, because it increases readability.
If only the items that are required are listed like this, you can see with a glance which parameters are required.
Same for default values. It's useful to easily identify the custom default values.
(Just like PEP8, this is purely from a readability point-of-view)

For the same reason we prefer lists to be alphabetically sorted, unless the order has a purpose.

Reply:
fixed